File: src/test/java/destiny/core/calendar/eightwords/EightWordsTest.java
/**
 * @author smallufo
 * Created on 2006/6/11 at 下午 11:23:55
 */
package destiny.core.calendar.eightwords;

import destiny.core.chinese.StemBranch;
import org.junit.Test;

import static org.junit.Assert.*;

public class EightWordsTest {

  @Test
  public void testEqualsSame() {
    EightWords expected = new EightWords("甲子", "甲子", "甲子", "甲子");
    EightWords actual = new EightWords(StemBranch.get("甲子"), StemBranch.get("甲子"), StemBranch.get("甲子"), StemBranch.get("甲子"));
    assertEquals(expected, actual);
    assertNotSame(expected, actual); //不應該 same
  }

  /** 測試傳入 null 到 EightWords 的 constructor 中 */
  @Test
  public void testNullInConstructor1_year() {
    try {
      new EightWords(null, StemBranch.get("乙丑"), StemBranch.get("丙寅"), StemBranch.get("丁卯"));
      fail("year's StemBranch is null , should throw RuntimeException !");
    } catch (RuntimeException expected) {
      assertTrue(true);
    }
  }

  @Test
  public void testNullInConstructor1_month() {
    try {
      new EightWords(StemBranch.get("甲子"), null, StemBranch.get("丙寅"), StemBranch.get("丁卯"));
      fail("month's StemBranch is null , should throw RuntimeException !");
    } catch (RuntimeException expected) {
      assertTrue(true);
    }
  }


  @Test
  public void testNullInConstructor1_day() {
    try {
      new EightWords(StemBranch.get("甲子"), StemBranch.get("乙丑"), null, StemBranch.get("丁卯"));
      fail("day's StemBranch is null , should throw RuntimeException !");
    } catch (Exception expected) {
      assertTrue(true);
    }
  }

  @Test
  public void testNullInConstructor1_hour() {
    try {

      new EightWords(StemBranch.get("甲子"), StemBranch.get("乙丑"), StemBranch.get("丙寅"), null);
      fail("day's StemBranch is null , should throw RuntimeException !");
    } catch (RuntimeException expected) {
      assertTrue(true);
    }
  }


  /** 測試傳入 null 到 EightWords 的 constructor 中 , 與前例不同在於 , 前例傳的是干支 , 本例傳的是字串 */
  @Test
  public void testNullInConstructor2() {
    try {
      new EightWords(null, "乙丑", "丙寅", "丁卯");
      fail("year's StemBranch is null , should throw RuntimeException !");
      new EightWords("甲子", null, "丙寅", "丁卯");
      fail("month's StemBranch is null , should throw RuntimeException !");
      new EightWords("甲子", "乙丑", null, "丁卯");
      fail("day's StemBranch is null , should throw RuntimeException !");
      new EightWords("甲子", "乙丑", "丙寅", null);
      fail("hour's StemBranch is null , should throw RuntimeException !");
    } catch (RuntimeException expected) {
      assertTrue(true);
    }
  }

  @Test
  public void testNullStemBranch() {
    EightWords eightWordsFull = new EightWords(StemBranch.get("甲子"), StemBranch.get("乙丑"), StemBranch.get("丙寅"), StemBranch.get("丁卯"));
    assertNotNull(eightWordsFull);
    assertNotNull(eightWordsFull.toString());
    System.out.println(eightWordsFull.toString());
  }
}

File: src/test/java/destiny/core/calendar/eightwords/EightWordsTest.kt
/**
 * @author smallufo
 * Created on 2006/6/11 at 下午 11:23:55
 */
package destiny.core.calendar.eightwords

import destiny.core.chinese.StemBranch
import destiny.core.chinese.StemBranchUnconstrained
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import mu.KotlinLogging
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNotSame

class EightWordsTest {

  val logger = KotlinLogging.logger { }

  @Test
  fun testSerializeMonthNormal() {
    val ew = EightWords("甲子", "乙丑", "丙寅", "丁卯")
    val encoded = Json.encodeToString(ew)
    logger.info { encoded }
    assertEquals("""{"year":"甲子","month":"乙丑","day":"丙寅","hour":"丁卯"}""", encoded)

    assertEquals(ew , Json.decodeFromString(encoded))
  }

  @Test
  fun testSerializeMonthUnconstrained() {
    val ew = EightWords(StemBranch.甲子, StemBranchUnconstrained.甲丑, StemBranch.丙寅, StemBranch.丁卯)
    val encoded = Json.encodeToString(ew)
    logger.info { encoded }
    assertEquals("""{"year":"甲子","month":"甲丑","day":"丙寅","hour":"丁卯"}""", encoded)
    assertEquals(ew, Json.decodeFromString(encoded))
  }

  @Test
  fun testEqualsSame() {
    val expected = EightWords("甲子", "甲子", "甲子", "甲子")
    val actual = EightWords(StemBranch["甲子"], StemBranch["甲子"], StemBranch["甲子"], StemBranch["甲子"])
    assertEquals(expected, actual)
    assertNotSame(expected, actual) //不應該 same
  }

  @Test
  fun testNullStemBranch() {
    val eightWordsFull = EightWords(StemBranch["甲子"], StemBranch["乙丑"], StemBranch["丙寅"], StemBranch["丁卯"])
    assertNotNull(eightWordsFull)
    assertNotNull(eightWordsFull.toString())
    println(eightWordsFull.toString())
  }

  @Test
  fun testStemBranches() {
    val eightWordsFull = EightWords(StemBranch["甲子"], StemBranch["乙丑"], StemBranch["丙寅"], StemBranch["丁卯"])
    assertEquals(listOf(StemBranch.甲子, StemBranch.乙丑, StemBranch.丙寅, StemBranch.丁卯), eightWordsFull.stemBranches)
  }
}
